cpp/oneapi/dal/detail/memory_impl_host.hpp

```diff
@@ -24,34 +24,48 @@
 namespace oneapi::dal::detail {
 namespace v1 {
 
-template <typename T>
-inline T* malloc(const default_host_policy&, std::int64_t count) {
-    return new T[count];
-}
+ONEDAL_EXPORT void* malloc(const default_host_policy&, std::size_t size);
+ONEDAL_EXPORT void* calloc(const default_host_policy&, std::size_t size);
+ONEDAL_EXPORT void free(const default_host_policy&, void* pointer);
+ONEDAL_EXPORT void fill(const default_host_policy&,
+                        void* dest,
+                        std::size_t size,
+                        const void* pattern,
+                        std::size_t pattern_size);
+ONEDAL_EXPORT void memset(const default_host_policy&,
+                          void* dest,
+                          std::int32_t value,
+                          std::int64_t size);
+ONEDAL_EXPORT void memcpy(const default_host_policy&,
+                          void* dest,
+                          const void* src,
+                          std::int64_t size);
 
 template <typename T>
-inline void free(const default_host_policy&, T* pointer) {
-    delete[] pointer;
+inline T* malloc(const default_host_policy& policy, std::int64_t count) {
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::size_t, sizeof(T), count);
+    const std::size_t bytes_count = sizeof(T) * count;
+    return static_cast<T*>(malloc(policy, bytes_count));
 }
 
-inline void memset(const default_host_policy&, void* dest, std::int32_t value, std::int64_t size) {
-    // TODO: is not safe since std::memset accepts size as size_t
-    // TODO: can be optimized in future
-    std::memset(dest, value, size);
+template <typename T>
+inline T* calloc(const default_host_policy& policy, std::int64_t count) {
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::size_t, sizeof(T), count);
+    const std::size_t bytes_count = sizeof(T) * count;
+    return static_cast<T*>(calloc(policy, bytes_count));
 }
 
-inline void memcpy(const default_host_policy&, void* dest, const void* src, std::int64_t size) {
-    // TODO: is not safe since std::memset accepts size as size_t
-    // TODO: can be optimized in future
-    std::memcpy(dest, src, size);
+template <typename T>
+inline void free(const default_host_policy& policy, T* pointer) {
+    using mutable_t = std::remove_const_t<T>;
+    free(policy, reinterpret_cast<void*>(const_cast<mutable_t*>(pointer)));
 }
 
 template <typename T>
-inline void fill(const default_host_policy&, T* dest, std::int64_t count, const T& value) {
-    // TODO: can be optimized in future
-    for (std::int64_t i = 0; i < count; i++) {
-        dest[i] = value;
-    }
+inline void fill(const default_host_policy& policy, T* dest, std::int64_t count, const T& value) {
```

Comment:
As we discussed, `array<T>` should work with any `T` incl. complex types. Not for all `T` you can just take address, cast to `std::uint8_t` and copy. Let's image `T` with non-trivial copy constructor, for example, it could be array of tables, strings or shared pointers.

Current implementation will lead to very tricky bugs. Imagine what happened to `std::shared_ptr` if we don't increment ref counter in new instance.

So my recommendation, is to keep `fill` in headers.

**For the future: we'll need to look through `array` implementation, I expect there are other implicit constraints on T**

Reply:
Thank you that you found that!

Reply:
> we'll need to look through array implementation

I think we can start by writing specific tests on `array` with non-trivial types of `T`